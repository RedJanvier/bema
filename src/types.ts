import BenchmarkJS from 'benchmark'
import { PrimedCase } from './execution/prime'
import { MaybePromise } from './utils'

/**
 * Summary data about a Bema run, all the groups, cases, etc..
 */
export type Report = {
  name: string
  time: TimeStats
  groups: GroupResultSimplified[]
}

export type ParameterName = string

export type BaseParameters = Record<ParameterName, string>

export type BaseContext = Record<string, unknown>

/** The context type in return postition. This forbids explicitly returning undefined. */
export type BaseContextReturnType<Returns extends BaseContext = BaseContext> = MaybePromise<void | Returns>

export type Provider<
  C extends BaseContext = BaseContext,
  P extends BaseParameters = BaseParameters,
  Returns extends BaseContext = BaseContext
> = {
  (context: BeforeCaseContext<C, P>, on: ContextProviderHooks<C, P>): BaseContextReturnType<Returns>
  parameters?: P
}

export type ContextProviderHooks<
  C extends BaseContext = BaseContext,
  P extends BaseParameters = BaseParameters
> = {
  /** Register a function to be run before the benchmark test is started. */
  before(callback: BeforeCaseCallback<C, P>): void
  /** Register a function to be run after the benchmark test has finished. */
  after(callback: AfterCaseCallback<C, P>): void
}

export type BeforeCaseCallback<
  C extends BaseContext = BaseContext,
  P extends BaseParameters = BaseParameters
> = (context: BeforeCaseContext<C, P>) => MaybePromise<void>

export type BeforeCaseContext<
  C extends BaseContext = BaseContext,
  P extends BaseParameters = BaseParameters
> = C & {
  $info: {
    parameters: P
  }
}

export type AfterCaseCallback<
  C extends BaseContext = BaseContext,
  P extends BaseParameters = BaseParameters,
  ReturnedValue = unknown
> = (context: AfterCaseContext<C, P, ReturnedValue>) => MaybePromise<void>

export type AfterCaseContext<
  C extends BaseContext = Record<string, never>,
  P extends BaseParameters = Record<string, never>,
  ReturnValue = unknown
> = C & {
  $info: CaseResult<P, ReturnValue>
}

export type CaseResult<P extends BaseParameters = BaseParameters, ReturnValue = unknown> = {
  /** Metadata about the case and stats about how it performed */
  report: CaseReport<P>
  /** The value returned by the last sample run. */
  returnValue?: ReturnValue
}

export type AfterGroupCallback<
  C extends BaseContext = BaseContext,
  P extends BaseParameters = BaseParameters
  // ReturnedValue = unknown
> = (context: AfterGroupContext<C, P>) => MaybePromise<void>

export type AfterGroupContext<
  C extends BaseContext = BaseContext,
  // TODO remove?
  _P extends BaseParameters = BaseParameters
> = C & {
  $info: GroupResult
}

export type CaseResultWithGroupStats = CaseResult & CaseReportGroupStats

type CaseReportGroupStats = {
  report: {
    stats: {
      /**
       * How much slower was this case than its peers in the group.
       *
       * Note that this value is scoped to the _matrix_ group it was run in.
       */
      percentSlower: number
    }
  }
}

export type CaseResultSimplifiedWithGroupStats = Omit<CaseResult, 'returnValue'> & CaseReportGroupStats

export type GroupResultSimplified = {
  /** Name of group. */
  name: string
  time: TimeStats
  caseResultsByMatrix: CaseResultSimplifiedWithGroupStats[][]
}

export type GroupResult = {
  /** Name of group. */
  name: string
  cases: PrimedCase[]
  casesSelected: PrimedCase[]
  caseResults: {
    byMatrix: Record<string, CaseResultWithGroupStats[]>
    all: CaseResultWithGroupStats[]
  }
  time: TimeStats
}

export type AfterBemaCallback<
  C extends BaseContext = BaseContext,
  P extends BaseParameters = BaseParameters
  // ReturnedValue = unknown
> = (context: AfterGroupContext<C, P>) => MaybePromise<void>

export type AfterBemaContext<
  C extends BaseContext = BaseContext,
  // TODO remove?
  _P extends BaseParameters = BaseParameters
> = C & {
  $info: BemaResult
}

export type BemaResult = {
  groupResults: GroupResult[]
}

export type ProviderReturnType<T, Passthrough> = T extends void
  ? Passthrough
  : Awaited<T> extends void
  ? Passthrough
  : T

export type CaseReport<P extends BaseParameters = BaseParameters> = {
  /**
   * Was the benchmark run in quick mode? If so this report is not statistically significant.
   */
  quickMode: boolean
  /** The name of the group that contained this benchmark. */
  group: string
  /** The name of this benchmark. */
  name: string
  /**
   * The matrix permutation under which this case was run.
   */
  matrix: BaseParameters[]
  /** The parameters this benchmark ran with */
  parameters: P
  parametersFromMatrix: BaseParameters
  parametersFromNotMatrix: BaseParameters
  /** That statistical results of how this benchmark performed. */
  stats: BenchmarkJS.Stats & {
    /** The arithmetic mean (aka. average) in milliseconds taken to run a sample. */
    mean: number
    /** On average how many times a bunchmark sample can run within one second. */
    hz: number
    /** The sample standard deviation in milliseconds */
    deviation: number
    /** The margin of error. */
    moe: number
    /** The relative margin of error (expressed as a percentage of the mean). */
    rme: number
    /** The standard error of the mean. */
    sem: number
    /** The sampled periods. */
    sample: number[]
    /** An object of timing data including cycle, elapsed, period, start, and stop. */
    times: BenchmarkJS.Times & {
      /** A timestamp of when the benchmark started (ms). */
      timeStamp: number
      /** The time taken to complete the benchmark (secs). */
      elapsed: number
      /** The time taken to execute the test once (secs). */
      period: number
      /** The time taken to complete the last cycle (secs). */
      cycle: number
    }
  }
}

export type ParameterChangeCallbackInfo = {
  activeCases: {
    parameters: Record<string, string>
  }[]
}

export type RegisteredParameterChangeCallback = {
  parameterName: string
  callback: ParameterChangeCallback
}

export type ParameterChangeCallback<ParameterName extends string = string> = (
  event: {
    name: ParameterName
    initial: boolean
    value: {
      before?: string
      after?: string
    }
  },
  info: ParameterChangeCallbackInfo
) => MaybePromise<void>

export type RunCallback<
  C extends BaseContext = BaseContext,
  P extends BaseParameters = BaseParameters,
  Returns = unknown
> = (context: BeforeCaseContext<C, P>) => MaybePromise<Returns>

export type TimeStats = {
  started: number
  finished: number
  /**
   * How long the entire section took to run in seconds.
   */
  elapsed: number
}

export type Awaited<T> = T extends PromiseLike<infer U> ? Awaited<U> : T
