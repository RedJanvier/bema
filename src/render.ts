import columnify from 'columnify'
import endent from 'endent'
import * as csv from 'fast-csv'
import kleur from 'kleur'
import markdownTable from 'markdown-table'
import { EOL } from 'os'
import slugify from 'slugify'
import { CaseReport, CaseResultSimplifiedWithGroupStats, GroupResultSimplified, Report } from './types'

export async function renderReportCSV(
  report: Report
): Promise<{ groupName: string; matrixName: string; content: string; name: string }[]> {
  return Promise.all(
    report.groups
      .flatMap((group) => {
        return Object.values(group.caseResultsByMatrix).map((caseResults) => ({
          groupName: group.name,
          caseResults,
        }))
      })
      .map(({ groupName, caseResults }) => {
        // @ts-expect-error FIXME
        const matrixName = caseResults[0].report.matrix
          .flatMap(Object.entries)
          .map((ent) => {
            return ent.map((s) => slugify(s, { lower: true })).join('=')
          })
          .join('+')
        return csv.writeToString(caseResults.map(getTabularData), { headers: true }).then((content) => ({
          groupName,
          matrixName,
          name: [slugify(groupName, { lower: true }), matrixName].filter((s) => s).join('--'),
          content,
        }))
      })
  )
}

export function renderReportMarkdown(report: Report): string {
  const title = `# Benchmark Report: ${report.name}`
  const groups = report.groups
    .map(
      (group) => endent`
              ## ${group.name}

              ${renderGroupResultForMarkdown(group)}
            `
    )
    .join(`${EOL + EOL}<br/>${EOL + EOL}`)

  return endent`
      ${title}

      ${groups}
    `
}

export function renderGroupResultForMarkdown(groupResult: GroupResultSimplified): string {
  return groupResult.caseResultsByMatrix
    .map((matrixCaseResults) => {
      // @ts-expect-error FIXME
      const matrixName = Object.values(matrixCaseResults[0].report.parametersFromMatrix).join(' ')
      const tabularData = matrixCaseResults.map(getTabularData)
      const table = markdownTable([Object.keys(tabularData[0] ?? {}), ...tabularData.map(Object.values)])
      const title = `MATRIX: ${matrixName.toUpperCase()}`
      return endent`
        ### ${title}

        ${table}
      `
    })
    .join(EOL + EOL)
}

export function renderGroupResultForTerminal(groupResult: GroupResultSimplified): string {
  return groupResult.caseResultsByMatrix
    .map((matrixCaseResults) => {
      // @ts-expect-error FIXME
      const matrixName = Object.values(matrixCaseResults[0].report.parametersFromMatrix).join(' ')
      const table = columnify(matrixCaseResults.map(getTabularData), { columnSplitter: '   ' })
      const title = kleur.bold(`MATRIX: ${matrixName.toUpperCase()}`)
      return `${title}${EOL}${table}`
    })
    .join(EOL + EOL)
}

export function getTabularData(
  caseResult: CaseResultSimplifiedWithGroupStats
): Record<string, string | number> {
  return {
    case: Object.values(caseResult.report.parametersFromNotMatrix).join(' '),
    'ops/s': renderHz(caseResult.report.stats.hz),
    '% slower than fastest': renderPercentSlowerThanFastest(caseResult.report.stats.percentSlower),
    'mean op duration (ms)': caseResult.report.stats.mean.toFixed(2),
    'standard deviation of mean op (ms)': caseResult.report.stats.deviation.toFixed(2),
    '% relative margin of error': `±${caseResult.report.stats.rme.toFixed(2)}%`,
    samples: caseResult.report.stats.sample.length,
  }
}

function renderPercentSlowerThanFastest(n: number): string {
  return n === 0 ? 'N/A' : `${n.toFixed(0)}%`
}

function renderHz(n: number) {
  return n > 50 ? Number(n.toFixed(0)) : Number(n.toFixed(2))
}

export function renderCaseSummary(report: CaseReport): string {
  const name = report.name
  const hz = renderHz(report.stats.hz)
  const rme = report.stats.rme.toFixed(2)
  const quick = report.quickMode ? ` (Quick Mode)` : ``
  const elapsed = report.stats.times.elapsed.toFixed(0)

  return endent`
    ${name} x ${kleur.green(hz)} ops/sec ±${rme}% ${kleur.dim(`${elapsed}s`)}${kleur.red(quick)}
  `
}
