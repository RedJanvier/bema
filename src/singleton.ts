import { create } from './declaration/bema'

export const bema = create()
