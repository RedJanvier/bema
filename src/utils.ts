import createDebug from 'debug'
import * as Path from 'path'
import { inspect } from 'util'

export const d = createDebug('bema')

export type MaybePromise<T> = T | Promise<T>

export type AdditionalPrismaStatsSamples = {
  prismaClientRaw: {
    engine: {
      timings: number[]
    }
  }
  prismaClient: {
    engine: {
      timings: number[]
    }
  }
}

export type AdditionalPrismaStats = {
  prismaClientRaw: {
    engine: {
      mean: number
      standardDeviation: number
    }
  }
  prismaClient: {
    engine: {
      mean: number
      standardDeviation: number
    }
  }
}

export function processPattern(pattern: string | RegExp): RegExp {
  return pattern instanceof RegExp ? pattern : new RegExp(`^${pattern}$`, 'i')
}

/** Exit with the given code and message. */
export function fatal(exitCode: number, message: string | Error): never
/** Exit with code 1 and the given message. */
export function fatal(message: string | Error): never
export function fatal(...args: [number, string | Error] | [string | Error]): never {
  let message
  let exitCode

  if (typeof args[0] === 'number') {
    exitCode = args[0]
    message = args[1]
  } else {
    exitCode = 1
    message = args[0]
  }

  console.error(message)
  process.exit(exitCode)
}

export function makeRelativePathExplicitlyRelative(path: string): string {
  if (Path.isAbsolute(path)) return path
  if (path.startsWith('./')) return path
  return `./${path}`
}

export function addExtensionIfNone(ext: string, path: string): string {
  if (/\.[^.]+$/.exec(path)) return path

  return path + '.' + ext.replace(/^\./, '')
}

//eslint-disable-next-line
export function arrayify<T = any>(
  x: T
  //eslint-disable-next-line
): Exclude<T, undefined> extends Array<any> ? Exclude<T, undefined> : Exclude<T, undefined>[] {
  // @ts-expect-error not sure
  if (Array.isArray(x)) return x
  // @ts-expect-error not sure
  if (x === undefined) return []
  // @ts-expect-error not sure
  return [x]
}

export type KeyNameToObjectWithKey<N extends string, T> = {
  [K in N as `${K}`]: T
}

export type OptionalKeys<T> = {
  [K in keyof T]?: T[K]
}

export function millisecondsToSeconds(n: number): number {
  return n / 1000
}

export function dump(...xs: unknown[]): void {
  console.log(...xs.map((x) => inspect(x, { depth: 20 })))
}

// TODO can this become Record<string, never>?
// eslint-disable-next-line
export type StringKeyOfOrString<T> = {} extends T ? string : Exclude<keyof T, number | symbol>

export function assertCasesHandled(x: never): void {
  throw new Error(`Unhandled casee: ${inspect(x)}`)
}
