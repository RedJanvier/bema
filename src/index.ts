import { bema } from './singleton'

export { Bema, create } from './declaration/bema'
export * from './helpers'
export { bema }

export default bema
