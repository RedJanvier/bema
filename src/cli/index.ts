#!/usr/bin/env node

import { d, fatal } from '../utils'
import { LintError, UserError } from '../utils/errors'

d('cli starting')

process.on('unhandledRejection', (e: Error) => {
  fatal(e)
})

main().catch((e: Error) => {
  if (e instanceof LintError) {
    console.error(e.message)
    process.exit(3)
  }

  if (e instanceof UserError) {
    console.error(e.message)
    process.exit(2)
  }

  console.error(e)
  process.exit(1)
})

async function main(): Promise<void> {
  // eslint-disable-next-line
  const command = (
    process.argv[2] === 'report'
      ? require('./report')
      : process.argv[2] === 'ui'
      ? require('./ui')
      : process.argv[2] === 'bench'
      ? require('./bench')
      : // default command
        require('./bench')
  ).default as () => Promise<void>

  await command()
}
