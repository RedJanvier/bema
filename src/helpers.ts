import { BaseContext, BaseParameters, ContextProviderHooks } from './types'
import { MaybePromise } from './utils'

/**
 * Create a provider.
 *
 * Providers can be passed to the `.use` method.
 *
 * You can configure paramter values that any benchmark using your provider will get.
 *
 * About the provider:
 *
 * 1. Anything you return in your provider method will be shallowly merged into the context.
 *
 * 2. If your provider depends on some context (given by another provider) then type that in the `context`
 * provider parameter. Do not use the generics so that inference can do its job. There is a limitation of TS
 * thatforces this: https://github.com/microsoft/TypeScript/issues/26242#issuecomment-757571172
 *
 * 3. If you care about it you can use the `BeforeCaseContext` generic which gives you an accurate type for
 * context by adding types for the `$info` property. See example below.
 *
 * 4. If your provider needs to tap into the benchmark lifecycle then use the API in the second parameter
 * `on`.
 *
 * @example
 *
 *   // Simple
 *
 *   const foobar = createProvider({
 *     provider(context) {
 *       // All downstream providers, hooks, and benchmark will have access to this data:
 *       return {
 *         foo: 'bar',
 *       }
 *     },
 *   })
 *
 *   // Usage Examples
 *   // Place on root for all cases to have access
 *   bema.use(foobar)
 *   // Place on a specific group if only its cases need it
 *   bema.group('foo').use(foobar).case()
 *   // Place on a specific case if only needed there
 *   bema.case().use(foobar)
 *
 * @example
 *
 *   // Complex
 *
 *   const prisma = createProvider({
 *     // All cases using this provider will get this benchmark parameter arguments
 *     parameters: {
 *       orm: 'prisma',
 *     },
 *     // use BeforeCaseContext generic if you want access to the $info
 *     provider(context: BeforeCaseContext<DatabaseInfo>, on) {
 *       console.log('hey look at me: %j', context.$info.parameters)
 *
 *       const prisma: PrismaClient = new PrismaClient({
 *         datasources: {
 *           db: {
 *             url: context.databaseInfo.url,
 *           },
 *         },
 *       })
 *
 *       on.setup(async () => {
 *         await prisma.$connect()
 *       })
 *
 *       on.teardown(async () => {
 *         await prisma.$disconnect()
 *       })
 *
 *       return {
 *         db: prisma,
 *       }
 *     },
 *   })
 *
 */
export function createProvider<
  ContextRequired extends BaseContext,
  ContextContributed extends BaseContext
>(config: {
  /**
   * The parameter values that any benchmark using (or inheriting use of) this provider will get.
   *
   * @example
   *
   *   const qux = createProvider({
   *     parameters: {
   *       foo: 'bar',
   *     },
   *     //...
   *   })
   *
   *   const bench = benchmarker.parameter('foo')
   *
   *   bench.case().use(qux) // has value "bar" for parameter "foo"
   *
   */
  parameters?: BaseParameters
  /**
   * The provider implementation.
   *
   * @param context  The context passed down from any providers higher up.
   * @param on       API to hook into lifecycle events.
   * @returns A plain object that will be shallowly merged with context.
   */
  provider: (
    context: ContextRequired,
    on: ContextProviderHooks<ContextRequired>
    // on: ContextProviderHooks<ContextRequired & ContextContributed> // <-- Though its technically a more accurate typing doing this breaks return type inference.
  ) => MaybePromise<ContextContributed>
}): <Context extends ContextRequired>(
  context: Context,
  on: ContextProviderHooks
) => MaybePromise<Context & ContextContributed> {
  async function provider(ctx: Record<string, unknown>, on: unknown) {
    // @ts-expect-error investigate why ctx must be unknown
    const ctxContributed: Record<string, unknown> = await config.provider(ctx, on)

    const ctxNew = {
      ...ctx,
      ...ctxContributed,
    }

    return ctxNew
  }

  provider.parameters = config.parameters

  // @ts-expect-error investigate why ctx must be unknown
  return provider
}
