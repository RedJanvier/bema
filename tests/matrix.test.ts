import './__setup'
import { constant } from 'lodash'
import { createProvider } from '../src'

test('expands all cases below, each case receiving context and parameters given by the providers', async () => {
  const p1 = createProvider({ parameters: { level1: 'pp1' }, provider: constant({ level1: 'cp1' }) })
  const p2 = createProvider({ parameters: { level1: 'pp2' }, provider: constant({ level1: 'cp2' }) })
  let ga = bema.parameter('level1').parameter('c').matrix(p1, p2).group('a')
  ga.case({ c: '1' }).run(f1)
  await run(bema)
  expect(f1.mock.calls.length).toEqual(2)
  expect(f1.mock.calls[0][0]).toMatchObject({ level1: 'cp1' })
  expect(f1.mock.calls[1][0]).toMatchObject({ level1: 'cp2' })
  expect(f1.mock.calls[0][0]?.$info.parameters).toMatchObject({ level1: 'pp1', c: '1' })
  expect(f1.mock.calls[1][0]?.$info.parameters).toMatchObject({ level1: 'pp2', c: '1' })
})

test('nested matrixes expand one another and then all cases below, each case receiving parameters given by the providers', async () => {
  const p1 = createProvider({ parameters: { level1: 'pp1' }, provider: constant({ level1: 'cp1' }) })
  const p2 = createProvider({ parameters: { level1: 'pp2' }, provider: constant({ level1: 'cp2' }) })
  const p3 = createProvider({ parameters: { level2: 'pp3' }, provider: constant({ level2: 'cp3' }) })
  const p4 = createProvider({ parameters: { level2: 'pp4' }, provider: constant({ level2: 'cp4' }) })
  let ga = bema
    .parameter('level1')
    .parameter('level2')
    .parameter('c')
    .matrix(p1, p2)
    .matrix(p3, p4)
    .group('a')
  ga.case({ c: '1' }).run(f1)
  await run(bema)
  expect(f1.mock.calls.length).toEqual(4)
  expect(f1.mock.calls[0][0]).toMatchObject({ level1: 'cp1', level2: 'cp3' })
  expect(f1.mock.calls[1][0]).toMatchObject({ level1: 'cp1', level2: 'cp4' })
  expect(f1.mock.calls[2][0]).toMatchObject({ level1: 'cp2', level2: 'cp3' })
  expect(f1.mock.calls[3][0]).toMatchObject({ level1: 'cp2', level2: 'cp4' })
  expect(f1.mock.calls[0][0]?.$info.parameters).toMatchObject({ level1: 'pp1', level2: 'pp3', c: '1' })
  expect(f1.mock.calls[1][0]?.$info.parameters).toMatchObject({ level1: 'pp1', level2: 'pp4', c: '1' })
  expect(f1.mock.calls[2][0]?.$info.parameters).toMatchObject({ level1: 'pp2', level2: 'pp3', c: '1' })
  expect(f1.mock.calls[3][0]?.$info.parameters).toMatchObject({ level1: 'pp2', level2: 'pp4', c: '1' })
})

describe('context', () => {
  test('case can get context from provider from matrix', async () => {
    const p1 = createProvider({ parameters: { level1: 'pp1' }, provider: constant({ level1: 'cp1' }) })
    let ga = bema.parameter('level1').parameter('c').matrix(p1).group('a')
    ga.case({ c: '1' }).run(f1)
    await run(bema)
    expect(f1.mock.calls[0]).toMatchSnapshot()
  })
  test('case in group can get context from provider from matrix', async () => {
    const p1 = createProvider({ parameters: { level1: 'pp1' }, provider: constant({ level1: 'cp1' }) })
    let b = bema.parameter('level1').parameter('c').matrix(p1)
    b.group('foo').case({ c: '1' }).run(f1)
    await run(b)
    expect(f1.mock.calls[0]).toMatchSnapshot()
  })
})
