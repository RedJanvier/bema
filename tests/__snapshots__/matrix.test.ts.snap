// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context case can get context from provider from matrix 1`] = `
Array [
  Object {
    "$info": Object {
      "parameters": Object {
        "c": "1",
        "level1": "pp1",
      },
    },
    "level1": "cp1",
  },
]
`;

exports[`context case in group can get context from provider from matrix 1`] = `
Array [
  Object {
    "$info": Object {
      "parameters": Object {
        "c": "1",
        "level1": "pp1",
      },
    },
    "level1": "cp1",
  },
]
`;
